File: src/segment.jl
# The segment structure

mutable struct Segment
    p1::Vertex
    p2::Vertex
    checkInput::Function
    length::Function
    dim::Function

    
    function Segment(p1,p2)
        this = new()

        this.p1 = p1
        this.p2 = p2

        this.checkInput = function()
            return size(this.p1.v) == size(this.p2.v)       
        end

        this.length = function()
            return norm(this.p1-this.p2)
        end

        this.dim = function()
            if this.checkInput()
                return length(this.p1.v)
            else
                return false
            end
        end

        return this        
    end
end

function dotDist(p::Vector{<:Real}, segment::Segment) 
    if segment.checkInput()
        if length(p) == segment.dim()
            p1_p2 = segment.p1 -segment.p2
            p_p2 = p -segment.p2

            λ = LinearAlgebra.dot(p1_p2,p_p2)/LinearAlgebra.dot(p1_p2,p1_p2)
            λ = max(min(λ,1),0)

            p0 = λ*segment.p1 + (1-λ)*segment.p2 

            return LinearAlgebra.norm(p-p0)
        else
            return "dimention of p doesnt match dimention of segment"
        end
    else
        return "Segment has wrong dimentions"
    end
end

function dotDist(p::Vertex, segment::Segment) 
    if segment.checkInput()
        if length(p.v) == segment.dim()
            p1_p2 = segment.p1 -segment.p2
            p_p2 = p -segment.p2

            λ = LinearAlgebra.dot(p1_p2.v,p_p2.v)/LinearAlgebra.dot(p1_p2.v,p1_p2.v)
            λ = max(min(λ,1),0)

            p0 = λ*segment.p1 + (1-λ)*segment.p2 

            return LinearAlgebra.norm((p-p0).v)
        else
            return "dimention of p doesnt match dimention of segment"
        end
    else
        return "Segment has wrong dimentions"
    end
end

function xangle(seg::Segment)
    Δ = seg.p2-seg.p1
    flag=false
    if Δ[2] < 0
        Δ[2] = -Δ[2]
        flag = true
    end
    xang =atan(Δ[2],abs(Δ[1]))
    if Δ[1]<0
        xang = pi-xang
    end
    if flag
        xang=2*pi-xang
    end
    return xang

end

File: Bounds.py/segment.py
class Segment:
	def __init__(self, p1:Vertex, p2:Vertex):
		self.p1 = p1
		self.p2 = p2

	def checkInput(self):
		return len(self.p1.v) == len(self.p2.v)

	def length(self):
		if self.checkInput():
			return norm(subVertex(self.p1,self.p2))
		else:
			return False 

	def dim(self):
		if self.checkInput():
			return len(self.p1.v)
		else:
			return False

def dotDist(p:Vertex, segment:Segment):
	# Finds the minimal distance between a point p and
    # the line segment connecting points p1 and p2
	if segment.checkInput():
		if len(p.v) == segment.dim():
			p31 = subVertex(p, segment.p1)
			p21 = subvertex(segment.p2, segment.p1)

			t = dot(p31, p21)/dot(p21, p21)
			t = min(1,max(t,0))

			p0 = addVertex(segment.p1, lambdaVertex(t, p21))

			return norm(subVertex(p, p0))
		else:
			raise ValueError("dimension of p doesnt match dimension of segment")
	else:
		raise ValueError("segment has wrong dimensions")